File: validation/run_chunks.R
#!/usr/bin/env Rscript
# Extract the ```{r} chunks from mlb_statcast.qmd in order and execute them
# against a synthetic Statcast frame.
#
# Three things are stubbed so this runs without Baseball Savant:
#   * `library(baseballr)` is dropped -- the only thing the notebook uses it for
#     is the scrape, which is itself stubbed below
#   * the `scrape` chunk is replaced by a read of synthetic_statcast.csv
#   * the `setup` chunk's sampler settings are shrunk so the Stan fits finish in
#     minutes instead of hours (skip with FAST=0)
# Every other line executes exactly as written in the notebook.

args <- commandArgs(trailingOnly = TRUE)
QMD  <- if (length(args)) args[1] else "../mlb_statcast.qmd"
FAST <- Sys.getenv("FAST", "1") == "1"

txt <- readLines(QMD, warn = FALSE)
starts <- grep("^```\\{r\\}", txt)
ends   <- grep("^```\\s*$", txt)

chunks <- lapply(starts, function(s) {
  e <- ends[ends > s][1]
  body <- txt[(s + 1):(e - 1)]
  lbl <- grep("^#\\|\\s*label:", body, value = TRUE)
  list(
    label = if (length(lbl)) sub("^#\\|\\s*label:\\s*", "", lbl[1]) else "?",
    code  = body[!grepl("^#\\|", body)]
  )
})
cat(sprintf("extracted %d R chunks\n", length(chunks)))

env <- new.env(parent = globalenv())
failures <- character(0)

for (i in seq_along(chunks)) {
  ch <- chunks[[i]]
  cat(sprintf("\n%s\n[%02d] %s  (%s)\n%s\n",
              strrep("=", 70), i - 1L, ch$label, format(Sys.time(), "%H:%M:%S"),
              strrep("=", 70)))
  flush(stdout())

  code <- ch$code
  if (ch$label == "setup") {
    code <- grep("^\\s*library\\(baseballr\\)", code, value = TRUE, invert = TRUE)
    cat(">> STUB: dropped library(baseballr); the scrape is stubbed too\n")
  }
  if (ch$label == "scrape") {
    code <- 'data <- readr::read_csv("synthetic_statcast.csv", show_col_types = FALSE); nrow(data)'
    cat(">> STUB: reading synthetic_statcast.csv instead of scraping\n")
  }

  res <- try(eval(parse(text = paste(code, collapse = "\n")), envir = env),
             silent = TRUE)
  if (inherits(res, "try-error")) {
    cat("FAILED:\n"); cat(as.character(res), "\n")
    failures <- c(failures, sprintf("[%02d] %s", i - 1L, ch$label))
    break
  } else if (!is.null(res) && ch$label != "scrape") {
    if (is.data.frame(res) || is.atomic(res)) try(print(res), silent = TRUE)
  }

  if (ch$label == "setup" && FAST) {
    assign("CHAINS", 2L, envir = env)
    assign("ITER",   600L, envir = env)
    assign("WARMUP", 400L, envir = env)
    cat(">> VALIDATION OVERRIDE: CHAINS=2 ITER=600 WARMUP=400\n")
  }
}

cat("\n", strrep("=", 70), "\n", sep = "")
cat("FAILURES:", if (length(failures)) paste(failures, collapse = ", ") else "none", "\n")
quit(status = if (length(failures)) 1L else 0L)

File: validation/run_chunks.py
"""Extract the ```{python} chunks from mlb_statcast.qmd in order and execute
them against a synthetic pitcher_summary.csv.

Only one thing is overridden: SAMPLE_KW is shrunk so the four MCMC fits finish
in minutes instead of hours. Every other line runs exactly as written in the
notebook.
"""
import re, sys, os, traceback
import matplotlib
matplotlib.use("Agg")

QMD = sys.argv[1] if len(sys.argv) > 1 else "../mlb_statcast.qmd"
FAST = os.environ.get("FAST", "1") == "1"

text = open(QMD).read()
chunks = re.findall(r"^```\{python\}\n(.*?)^```", text, re.S | re.M)
print(f"extracted {len(chunks)} python chunks", flush=True)

def label_of(src):
    m = re.search(r"#\|\s*label:\s*(\S+)", src)
    return m.group(1) if m else "?"

def strip_opts(src):
    return "\n".join(l for l in src.splitlines() if not l.strip().startswith("#|"))

g = {"__name__": "__main__"}
failures = []
for i, raw in enumerate(chunks):
    lbl = label_of(raw)
    code = strip_opts(raw)
    import time as _t
    print(f"\n{'='*70}\n[{i:02d}] {lbl}  ({_t.strftime('%H:%M:%S')})\n{'='*70}", flush=True)
    try:
        exec(compile(code, f"<chunk:{lbl}>", "exec"), g)
    except Exception:
        traceback.print_exc()
        failures.append((i, lbl))
        break
    if lbl == "py-setup" and FAST:
        g["SAMPLE_KW"].update(draws=200, tune=300, chains=2)
        print(">> VALIDATION OVERRIDE:", g["SAMPLE_KW"], flush=True)

print("\n" + "=" * 70)
print("FAILURES:", failures if failures else "none")
sys.exit(1 if failures else 0)
